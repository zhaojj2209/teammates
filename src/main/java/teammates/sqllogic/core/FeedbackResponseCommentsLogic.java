package teammates.sqllogic.core;

import java.util.UUID;

import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.sqlapi.FeedbackResponseCommentsDb;
import teammates.storage.sqlentity.FeedbackResponseComment;
import teammates.ui.request.FeedbackResponseCommentUpdateRequest;

/**
 * Handles operations related to feedback response comments.
 *
 * @see FeedbackResponseComment
 * @see FeedbackResponseCommentsDb
 */
public final class FeedbackResponseCommentsLogic {

    private static final FeedbackResponseCommentsLogic instance = new FeedbackResponseCommentsLogic();
    private FeedbackResponseCommentsDb frcDb;

    private FeedbackResponseCommentsLogic() {
        // prevent initialization
    }

    public static FeedbackResponseCommentsLogic inst() {
        return instance;
    }

    /**
     * Initialize dependencies for {@code FeedbackResponseCommentsLogic}.
     */
    void initLogicDependencies(FeedbackResponseCommentsDb frcDb) {
        this.frcDb = frcDb;
    }

    /**
     * Gets an feedback response comment by feedback response comment id.
     * @param id of feedback response comment.
     * @return the specified feedback response comment.
     */
    public FeedbackResponseComment getFeedbackResponseComment(Long id) {
        return frcDb.getFeedbackResponseComment(id);
    }

    /**
     * Gets the comment associated with the response.
     */
    public FeedbackResponseComment getFeedbackResponseCommentForResponseFromParticipant(
            UUID feedbackResponseId) {
        return frcDb.getFeedbackResponseCommentForResponseFromParticipant(feedbackResponseId);
    }

    /**
     * Creates a feedback response comment.
     * @throws EntityAlreadyExistsException if the comment alreadty exists
     * @throws InvalidParametersException if the comment is invalid
     */
    public FeedbackResponseComment createFeedbackResponseComment(FeedbackResponseComment frc)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return frcDb.createFeedbackResponseComment(frc);
    }

    /**
     * Deletes a feedbackResponseComment.
     */
    public void deleteFeedbackResponseComment(Long frcId) {
        frcDb.deleteFeedbackResponseComment(frcId);
    }

    /**
     * Updates a feedback response comment.
     * @throws EntityDoesNotExistException if the comment does not exist
     */
    public FeedbackResponseComment updateFeedbackResponseComment(Long frcId,
            FeedbackResponseCommentUpdateRequest updateRequest, String updaterEmail)
            throws EntityDoesNotExistException {
        FeedbackResponseComment comment = frcDb.getFeedbackResponseComment(frcId);
        if (comment == null) {
            throw new EntityDoesNotExistException("Trying to update a feedback response comment that does not exist.");
        }

        comment.setCommentText(updateRequest.getCommentText());
        comment.setShowCommentTo(updateRequest.getShowCommentTo());
        comment.setShowGiverNameTo(updateRequest.getShowGiverNameTo());
        comment.setLastEditorEmail(updaterEmail);

        return comment;
    }

    /**
     * Updates all email fields of feedback response comments.
     */
    public void updateFeedbackResponseCommentsEmails(String courseId, String oldEmail, String updatedEmail) {
        frcDb.updateGiverEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
        frcDb.updateLastEditorEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
    }

}
