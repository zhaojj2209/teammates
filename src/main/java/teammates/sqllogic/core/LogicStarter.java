package teammates.sqllogic.core;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;

import teammates.common.util.Logger;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlapi.CoursesDb;
import teammates.storage.sqlapi.DeadlineExtensionsDb;
import teammates.storage.sqlapi.FeedbackQuestionsDb;
import teammates.storage.sqlapi.FeedbackResponseCommentsDb;
import teammates.storage.sqlapi.FeedbackResponsesDb;
import teammates.storage.sqlapi.FeedbackSessionsDb;
import teammates.storage.sqlapi.NotificationsDb;
import teammates.storage.sqlapi.UsageStatisticsDb;
import teammates.storage.sqlapi.UsersDb;

/**
 * Setup in web.xml to register logic classes at application startup.
 */
public class LogicStarter implements ServletContextListener {

    private static final Logger log = Logger.getLogger();

    /**
     * Registers dependencies between different logic classes.
     */
    public static void initializeDependencies() {
        AccountsLogic accountsLogic = AccountsLogic.inst();
        AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();
        CoursesLogic coursesLogic = CoursesLogic.inst();
        DataBundleLogic dataBundleLogic = DataBundleLogic.inst();
        DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
        FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
        FeedbackResponsesLogic frLogic = FeedbackResponsesLogic.inst();
        FeedbackResponseCommentsLogic frcLogic = FeedbackResponseCommentsLogic.inst();
        FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();
        NotificationsLogic notificationsLogic = NotificationsLogic.inst();
        UsageStatisticsLogic usageStatisticsLogic = UsageStatisticsLogic.inst();
        UsersLogic usersLogic = UsersLogic.inst();

        accountRequestsLogic.initLogicDependencies(AccountRequestsDb.inst());
        accountsLogic.initLogicDependencies(AccountsDb.inst(), notificationsLogic, usersLogic);
        coursesLogic.initLogicDependencies(CoursesDb.inst(), fsLogic, usersLogic);
        dataBundleLogic.initLogicDependencies(accountsLogic, accountRequestsLogic, coursesLogic,
                deadlineExtensionsLogic, fsLogic, fqLogic, frLogic, frcLogic,
                notificationsLogic, usersLogic);
        deadlineExtensionsLogic.initLogicDependencies(DeadlineExtensionsDb.inst(), fsLogic);
        fsLogic.initLogicDependencies(FeedbackSessionsDb.inst(), coursesLogic, frLogic, fqLogic);
        frLogic.initLogicDependencies(FeedbackResponsesDb.inst(), usersLogic, fqLogic);
        frcLogic.initLogicDependencies(FeedbackResponseCommentsDb.inst());
        fqLogic.initLogicDependencies(FeedbackQuestionsDb.inst(), coursesLogic, frLogic, usersLogic, fsLogic);
        notificationsLogic.initLogicDependencies(NotificationsDb.inst());
        usageStatisticsLogic.initLogicDependencies(UsageStatisticsDb.inst());
        usersLogic.initLogicDependencies(UsersDb.inst(), accountsLogic, frLogic, frcLogic, deadlineExtensionsLogic);
        log.info("Initialized dependencies between logic classes");
    }

    @Override
    public void contextInitialized(ServletContextEvent event) {
        // Invoked by Jetty at application startup.
        initializeDependencies();
    }

    @Override
    public void contextDestroyed(ServletContextEvent event) {
        // Nothing to do
    }

}
