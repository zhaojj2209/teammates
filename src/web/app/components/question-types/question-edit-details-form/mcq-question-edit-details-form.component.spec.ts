import { DragDropModule } from '@angular/cdk/drag-drop';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { TeammatesCommonModule } from '../../teammates-common/teammates-common.module';
import { McqFieldComponent } from './mcq-field/mcq-field.component';
import { McqQuestionEditDetailsFormComponent } from './mcq-question-edit-details-form.component';
import { WeightFieldComponent } from './weight-field/weight-field.component';

describe('McqQuestionEditDetailsFormComponent', () => {
  let component: McqQuestionEditDetailsFormComponent;
  let fixture: ComponentFixture<McqQuestionEditDetailsFormComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [
        McqQuestionEditDetailsFormComponent,
        McqFieldComponent,
        WeightFieldComponent,
      ],
      imports: [
        FormsModule,
        DragDropModule,
        TeammatesCommonModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(McqQuestionEditDetailsFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view i.e. Question options are radio buttons', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when questionDropdownEnabled is true', () => {
    component.triggerQuestionDropdownEnabled(true);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should snap when questionDropdownEnabled is false', () => {
    component.triggerQuestionDropdownEnabled(false);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should be false by default when checkbox to show options as a drop-down is unchecked', () => {
    const dropDownCheckBoxNe: any = fixture.debugElement.nativeElement.querySelector('#make-question-dropdown');

    expect(component.model.questionDropdownEnabled).toBeFalsy();
    expect(dropDownCheckBoxNe.checked).toBeFalsy();
  });
});
