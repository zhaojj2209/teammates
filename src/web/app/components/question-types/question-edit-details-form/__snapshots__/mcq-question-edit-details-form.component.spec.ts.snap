// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`McqQuestionEditDetailsFormComponent should snap when questionDropdownEnabled is false 1`] = `
<tm-mcq-question-edit-details-form
  PARTICIPANT_TYPES={[Function Array]}
  detailsChange={[Function EventEmitter_]}
  isEditable={[Function Boolean]}
  model={[Function Object]}
  storageModel={[Function Object]}
>
  <div
    class="row"
  >
    <br />
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="bold-label"
        >
          <input
            checked="model.questionDropdownEnabled"
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="make-question-dropdown"
            type="checkbox"
          />
          <b>
            Show options as a drop-down
          </b>
        </label>
      </div>
    </div>
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="form-check-label"
          container="body"
          ngbtooltip="Assign weights to the choices for calculating statistics."
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="weights-checkbox"
            type="checkbox"
          />
          <b>
            Options are weighted
          </b>
        </label>
      </div>
    </div>
    <br />
    <div
      class="row col-sm-12 row-cols-2 mt-4"
    >
      <div
        class="col-sm-6"
      >
        <div>
          <div
            cdkdroplist=""
            class="cdk-drop-list"
            id="options-section"
          >
            <br />
          </div>
          <div
            class="row"
          >
            <div
              class="col-sm-6"
            >
              <button
                class="btn btn-primary"
                id="btn-add-option"
                name="button"
                ngdefaultcontrol=""
              >
                <i
                  class="fas fa-plus"
                />
                 add more options 
              </button>
            </div>
          </div>
          <br />
          <div
            class="row"
          >
            <div
              class="checkbox col-sm-9 padding-left-35px"
            >
              <label
                class="bold-label"
              >
                <input
                  class="form-check-input ng-untouched ng-pristine ng-valid"
                  id="other-checkbox"
                  type="checkbox"
                />
                 Add 'Other' option (Allows respondents to type in their own answer) 
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-sm-6"
      >
        <label
          class="form-check-label mb-2"
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="generate-checkbox"
            type="checkbox"
          />
          <b>
            Or, generate options from the list of all
          </b>
        </label>
        <select
          class="form-control ng-untouched ng-pristine"
          disabled=""
          id="generate-dropdown"
        >
          <option
            value="STUDENTS"
          >
            students
          </option>
          <option
            value="STUDENTS_EXCLUDING_SELF"
          >
            students (excluding self)
          </option>
          <option
            value="TEAMS"
          >
            teams
          </option>
          <option
            value="TEAMS_EXCLUDING_SELF"
          >
            teams (excluding own team)
          </option>
          <option
            value="OWN_TEAM_MEMBERS_INCLUDING_SELF"
          >
            team members
          </option>
          <option
            value="OWN_TEAM_MEMBERS"
          >
            team members (excluding self)
          </option>
          <option
            value="INSTRUCTORS"
          >
            instructors
          </option>
        </select>
      </div>
    </div>
  </div>
</tm-mcq-question-edit-details-form>
`;

exports[`McqQuestionEditDetailsFormComponent should snap when questionDropdownEnabled is true 1`] = `
<tm-mcq-question-edit-details-form
  PARTICIPANT_TYPES={[Function Array]}
  detailsChange={[Function EventEmitter_]}
  isEditable={[Function Boolean]}
  model={[Function Object]}
  storageModel={[Function Object]}
>
  <div
    class="row"
  >
    <br />
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="bold-label"
        >
          <input
            checked="model.questionDropdownEnabled"
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="make-question-dropdown"
            type="checkbox"
          />
          <b>
            Show options as a drop-down
          </b>
        </label>
      </div>
    </div>
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="form-check-label"
          container="body"
          ngbtooltip="Assign weights to the choices for calculating statistics."
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="weights-checkbox"
            type="checkbox"
          />
          <b>
            Options are weighted
          </b>
        </label>
      </div>
    </div>
    <br />
    <div
      class="row col-sm-12 row-cols-2 mt-4"
    >
      <div
        class="col-sm-6"
      >
        <div>
          <div
            cdkdroplist=""
            class="cdk-drop-list"
            id="options-section"
          >
            <br />
          </div>
          <div
            class="row"
          >
            <div
              class="col-sm-6"
            >
              <button
                class="btn btn-primary"
                id="btn-add-option"
                name="button"
                ngdefaultcontrol=""
              >
                <i
                  class="fas fa-plus"
                />
                 add more options 
              </button>
            </div>
          </div>
          <br />
          <div
            class="row"
          >
            <div
              class="checkbox col-sm-9 padding-left-35px"
            >
              <label
                class="bold-label"
              >
                <input
                  class="form-check-input ng-untouched ng-pristine ng-valid"
                  id="other-checkbox"
                  type="checkbox"
                />
                 Add 'Other' option (Allows respondents to type in their own answer) 
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-sm-6"
      >
        <label
          class="form-check-label mb-2"
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="generate-checkbox"
            type="checkbox"
          />
          <b>
            Or, generate options from the list of all
          </b>
        </label>
        <select
          class="form-control ng-untouched ng-pristine"
          disabled=""
          id="generate-dropdown"
        >
          <option
            value="STUDENTS"
          >
            students
          </option>
          <option
            value="STUDENTS_EXCLUDING_SELF"
          >
            students (excluding self)
          </option>
          <option
            value="TEAMS"
          >
            teams
          </option>
          <option
            value="TEAMS_EXCLUDING_SELF"
          >
            teams (excluding own team)
          </option>
          <option
            value="OWN_TEAM_MEMBERS_INCLUDING_SELF"
          >
            team members
          </option>
          <option
            value="OWN_TEAM_MEMBERS"
          >
            team members (excluding self)
          </option>
          <option
            value="INSTRUCTORS"
          >
            instructors
          </option>
        </select>
      </div>
    </div>
  </div>
</tm-mcq-question-edit-details-form>
`;

exports[`McqQuestionEditDetailsFormComponent should snap with default view i.e. Question options are radio buttons 1`] = `
<tm-mcq-question-edit-details-form
  PARTICIPANT_TYPES={[Function Array]}
  detailsChange={[Function EventEmitter_]}
  isEditable={[Function Boolean]}
  model={[Function Object]}
  storageModel={[Function Object]}
>
  <div
    class="row"
  >
    <br />
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="bold-label"
        >
          <input
            checked="model.questionDropdownEnabled"
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="make-question-dropdown"
            type="checkbox"
          />
          <b>
            Show options as a drop-down
          </b>
        </label>
      </div>
    </div>
    <div
      class="col-sm-12"
    >
      <div
        class="form-check"
      >
        <label
          class="form-check-label"
          container="body"
          ngbtooltip="Assign weights to the choices for calculating statistics."
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="weights-checkbox"
            type="checkbox"
          />
          <b>
            Options are weighted
          </b>
        </label>
      </div>
    </div>
    <br />
    <div
      class="row col-sm-12 row-cols-2 mt-4"
    >
      <div
        class="col-sm-6"
      >
        <div>
          <div
            cdkdroplist=""
            class="cdk-drop-list"
            id="options-section"
          >
            <br />
          </div>
          <div
            class="row"
          >
            <div
              class="col-sm-6"
            >
              <button
                class="btn btn-primary"
                id="btn-add-option"
                name="button"
                ngdefaultcontrol=""
              >
                <i
                  class="fas fa-plus"
                />
                 add more options 
              </button>
            </div>
          </div>
          <br />
          <div
            class="row"
          >
            <div
              class="checkbox col-sm-9 padding-left-35px"
            >
              <label
                class="bold-label"
              >
                <input
                  class="form-check-input ng-untouched ng-pristine ng-valid"
                  id="other-checkbox"
                  type="checkbox"
                />
                 Add 'Other' option (Allows respondents to type in their own answer) 
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-sm-6"
      >
        <label
          class="form-check-label mb-2"
        >
          <input
            class="form-check-input ng-untouched ng-pristine ng-valid"
            id="generate-checkbox"
            type="checkbox"
          />
          <b>
            Or, generate options from the list of all
          </b>
        </label>
        <select
          class="form-control ng-untouched ng-pristine ng-valid"
          disabled=""
          id="generate-dropdown"
        >
          <option
            value="STUDENTS"
          >
            students
          </option>
          <option
            value="STUDENTS_EXCLUDING_SELF"
          >
            students (excluding self)
          </option>
          <option
            value="TEAMS"
          >
            teams
          </option>
          <option
            value="TEAMS_EXCLUDING_SELF"
          >
            teams (excluding own team)
          </option>
          <option
            value="OWN_TEAM_MEMBERS_INCLUDING_SELF"
          >
            team members
          </option>
          <option
            value="OWN_TEAM_MEMBERS"
          >
            team members (excluding self)
          </option>
          <option
            value="INSTRUCTORS"
          >
            instructors
          </option>
        </select>
      </div>
    </div>
  </div>
</tm-mcq-question-edit-details-form>
`;
